### PR TITLE
EKF2: improve and simplify yaw measurement jacobian

### PR DESCRIPTION
**Pull request by bresch**
### Solved Problem
The current derivation for the yaw fusion uses Euler angles and required two different rotation sequences to avoid gimbal lock.

### Solution
A yaw perturbation can be defined around the vertical axis of the tangent space of a global rotation error. This definition is exact, does not require trigonometric functions as small angle approximations can be safely used for perturbations and is gimbal-lock-free.

Furthermore, we expect the yaw innovation variance to be the sum of the state yaw variance and the measurement noise but looking at the result of the current implementation, we can see that there are some -small- inconsistencies (note that the measurement variance -a constant value- was removed in the plot below to be able to compare the results):
Screenshot from 2024-03-13 11-17-32

With the new derivation the innovation variance obtained using the measurement jacobian matches precisely the value obtained from the state yaw variance:
Screenshot from 2024-03-13 11-20-32

Log: https://logs.px4.io/plot_app?log=2883d849-a8eb-4b74-81d1-aa0e841efa2f
Note that for this log I forced the mag heading fusion to run on every sample (not only when the innovation is too large).

### Changelog Entry
For release notes:
```
EKF2: improve and simplify yaw measurement jacobian
New parameter: -
Documentation: -
```

### Test coverage
unit test, replay, SITL

### CODE REVIEW

File: src/modules/ekf2/EKF/python/ekf_derivation/derivation.py [hunk 1]
```diff
@@ -390,6 +390,24 @@ def compute_mag_z_innov_var_and_h(
 
     return (innov_var, H.T)
 
+def compute_yaw_innov_var_and_h(
```
Comment (dagar) [inline src/modules/ekf2/EKF/python/ekf_derivation/derivation.py:393]: Would it overall be simpler if we just computed the innovation in here as well?

Comment (bresch) [reply src/modules/ekf2/EKF/python/ekf_derivation/derivation.py:393]: I don't think there will be much computational benefit in doing so; it's just a couple of multiplications and additions to get the yaw innovation from 2 quaternions

**Review by dagar — APPROVED** [review]

**Comment by priseborough** [comment]
We might need a check for the max allowable tilt misalignment between state and observation quaternion.

**Comment by bresch** [comment]
> We might need a check for the max allowable tilt misalignment between state and observation quaternion.

Yes, we can do that if we change the computation of the "yaw innovation" using quaternions directly.

**Pull request merged**